File: scripts/generate-permanent-tasks.py
#!/usr/bin/env python3
"""
Generate Permanent Conversion Tasks (Python version)

Scan literature/ directory and output JSONL tasks for permanent-converter.
Fixes path/encoding issues common in Node.js.
"""

import json
import os
import sys
import re
from pathlib import Path
from datetime import datetime

PROJECT_DIR = Path('/Users/yuxiang/workspaces/my_openclaw/.openclaw/workspace')
LITERATURE_DIR = PROJECT_DIR / 'projects' / 'ai-blog' / 'zettelkasten' / 'literature'
TENTATIVE_DIR = LITERATURE_DIR / 'tentative'
PERMANENT_DIR = PROJECT_DIR / 'projects' / 'ai-blog' / 'zettelkasten' / 'permanent'
TASKS_DIR = PROJECT_DIR / 'projects' / 'ai-blog' / 'scripts' / 'tasks'

# Ensure directories exist
TASKS_DIR.mkdir(parents=True, exist_ok=True)
TENTATIVE_DIR.mkdir(parents=True, exist_ok=True)

print('🔍 Scanning literature notes for permanent conversion...\n')

# Collect all literature files (including tentative)
lit_files = []

# From main literature dir
if LITERATURE_DIR.exists():
    for f in LITERATURE_DIR.iterdir():
        if f.is_file() and f.suffix == '.md' and f.name != 'README.md':
            lit_files.append({'file': f.name, 'tentative': False})

# From tentative dir
if TENTATIVE_DIR.exists():
    for f in TENTATIVE_DIR.iterdir():
        if f.is_file() and f.suffix == '.md':
            lit_files.append({'file': f.name, 'tentative': True})

print(f'   Found {len(lit_files)} literature notes')

# Read existing permanent sources to dedupe
permanent_sources = set()
if PERMANENT_DIR.exists():
    for f in PERMANENT_DIR.glob('*.md'):
        try:
            content = f.read_text(encoding='utf-8')
            # Look for literature_source: "filename.md"
            match = re.search(r'literature_source:\s*"([^"]+)"', content)
            if match:
                permanent_sources.add(match.group(1))
        except Exception as e:
            print(f'   ⚠️  Skipping {f.name}: {e}')

# Filter pending tasks
tasks = []
for item in lit_files:
    if item['file'] not in permanent_sources:
        rel_path = 'projects/ai-blog/zettelkasten/literature/tentative' if item['tentative'] else 'projects/ai-blog/zettelkasten/literature'
        base_dir = TENTATIVE_DIR if item['tentative'] else LITERATURE_DIR
        tasks.append({
            'file': f'{rel_path}/{item["file"]}',
            'tentative': item['tentative'],
            'timestamp': os.path.getmtime(str(base_dir / item['file']))
        })

print(f'   Pending conversion: {len(tasks)} notes')

if not tasks:
    print('   🎉 All literature notes already converted to permanent!')
    sys.exit(0)

# Sort by timestamp (newest first)
tasks.sort(key=lambda x: x['timestamp'], reverse=True)

# Output JSONL
timestamp = datetime.now().strftime('%Y-%m-%dT%H-%M-%S').replace(':', '-')
out_file = TASKS_DIR / f'permanent-convert-{timestamp}.jsonl'

with out_file.open('w', encoding='utf-8') as f:
    for t in tasks:
        f.write(json.dumps(t, ensure_ascii=False) + '\n')

print(f'\n📦 Task file generated: {out_file.name}')
print(f'\n📋 Summary:')
print(f'   Total literature notes: {len(lit_files)}')
print(f'   Already converted: {len(lit_files) - len(tasks)}')
print(f'   Need conversion: {len(tasks)}')
print(f'   Tentative notes included: {sum(1 for t in tasks if t["tentative"])}')
print(f'\n🚀 To process:')
print(f'   python3 projects/ai-blog/scripts/generate-permanent-tasks.py')
print(f'   node scripts/permanent-convert-orchestrator-v3.js {out_file.name}')

File: scripts/generate-permanent-tasks.js
#!/usr/bin/env node
/**
 * Generate Permanent Conversion Tasks
 *
 * 扫描 literature/ 目录，找出所有未转换为 permanent 的笔记
 * 输出 JSONL 任务列表供 permanent-converter 处理
 */

const fs = require('fs');
const path = require('path');

const PROJECT_DIR = '/Users/yuxiang/workspaces/my_openclaw/.openclaw/workspace';
const LITERATURE_DIR = path.join(PROJECT_DIR, 'projects', 'ai-blog', 'zettelkasten', 'literature');
const TENTATIVE_DIR = path.join(LITERATURE_DIR, 'tentative');
const PERMANENT_DIR = path.join(PROJECT_DIR, 'projects', 'ai-blog', 'zettelkasten', 'permanent');
const TASKS_DIR = path.join(PROJECT_DIR, 'projects', 'ai-blog', 'scripts', 'tasks');

// 确保目录存在
if (!fs.existsSync(TASKS_DIR)) fs.mkdirSync(TASKS_DIR, { recursive: true });
if (!fs.existsSync(TENTATIVE_DIR)) fs.mkdirSync(TENTATIVE_DIR, { recursive: true });

console.log('🔍 Scanning literature notes for permanent conversion...\n');

// 收集所有literature文件（包括tentative）
const litFiles = [
  ...fs.readdirSync(LITERATURE_DIR).filter(f => f.endsWith('.md')).map(f => ({ file: f, tentative: false })),
  ...fs.readdirSync(TENTATIVE_DIR).filter(f => f.endsWith('.md')).map(f => ({ file: f, tentative: true }))
];

console.log(`   Found ${litFiles.length} literature notes`);

// 读取已有permanent notes的source记录
const permanentSources = new Set();
fs.readdirSync(PERMANENT_DIR).filter(f => f.endsWith('.md')).forEach(f => {
  const content = fs.readFileSync(path.join(PERMANENT_DIR, f), 'utf8');
  const match = content.match(/literature_source:\s*"([^"]+)"/);
  if (match) {
    permanentSources.add(match[1]);
  }
});

// 过滤出未转换的
const tasks = [];
litFiles.forEach(({ file, tentative }) => {
  if (!permanentSources.has(file)) {
    tasks.push({
      file: path.join(tentative ? 'projects/ai-blog/zettelkasten/literature/tentative' : 'projects/ai-blog/zettelkasten/literature', file),
      tentative,
      timestamp: fs.statSync(path.join(tentative ? TENTATIVE_DIR : LITERATURE_DIR, file)).mtimeMs
    });
  }
});

console.log(`   Pending conversion: ${tasks.length} notes`);

if (tasks.length === 0) {
  console.log('   🎉 All literature notes already converted to permanent!');
  process.exit(0);
}

// 按时间排序（优先转换最新的）
tasks.sort((a, b) => b.timestamp - a.timestamp);

// 输出 JSONL
const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
const outFile = path.join(TASKS_DIR, `permanent-convert-${timestamp}.jsonl`);
const jsonl = tasks.map(t => JSON.stringify(t)).join('\n');
fs.writeFileSync(outFile, jsonl, 'utf8');

console.log(`\n📦 Task file generated: ${path.basename(outFile)}`);
console.log(`\n📋 Summary:`);
console.log(`   Total literature notes: ${litFiles.length}`);
console.log(`   Already converted: ${litFiles.length - tasks.length}`);
console.log(`   Need conversion: ${tasks.length}`);
console.log(`   Tentative notes included: ${tasks.filter(t => t.tentative).length}`);
console.log(`\n🚀 To process:`);
console.log(`   node scripts/permanent-convert-orchestrator-v3.js ${path.basename(outFile)}`);
console.log(`   (or integrate into heartbeat)`);
